### DISCUSSION TITLE
How does pydantic implement `__eq__` and `__hash__`?

### DISCUSSION THREAD

**Discussion by Ravencentric** [pydantic/pydantic#9867] [Question] [answered]
I could not find anything concrete that covered how exactly pydantic implements `__eq__` and `__hash__` methods in the documentation. So I did some quick tests and seems like it's based on every attribute being equal?

```py
>>> from pydantic import BaseModel, ConfigDict
>>>
>>> class Test(BaseModel):
...     model_config = ConfigDict(frozen=True)
...     x: int | None = None
...     y: str | None = None
>>> a = Test(x=1, y="a")
>>> b = Test(x=1, y="a")
>>> c = Test(x=1, y="b")
>>> d = Test()
>>> a == b
True

>>> a == c
False

>>> set((a, b, c, d))
{Test(x=1, y='a'), Test(x=1, y='b'), Test(x=None, y=None)}
```

**Comment by uriyyo** [comment, marked as answer]
Hi @Ravencentric,

You can find the implementation here:

https://github.com/pydantic/pydantic/blob/cc90fb060d9f7b9abb7cecafc118a06f2b5df4f9/pydantic/main.py#L931-L984

**Reply by Ravencentric** [reply]
Thank you
